ocsp.go

```diff
@@ -407,101 +515,212 @@ func verifyPeerCertificate(callback func(*sync.WaitGroup, []*x509.Certificate) [
+			if r.err != nil && (r.code == ocspRevoked || ocspFailOpen == ocspFailOpenFalse) {
+				// return error if revoked or OCSPFailOpen mode is false
 				return r.err
 			}
 		}
 	}
-	writeOCSPCacheFile()
+
+	ocspResponseCacheLock.Lock()
+	if cacheUpdated {
+		writeOCSPCacheFile()
+	}
+	cacheUpdated = false
+	ocspResponseCacheLock.Unlock()
 	return nil
 }
 
-func getAllRevocationStatusParallel(wg *sync.WaitGroup, verifiedChains []*x509.Certificate) []*ocspStatus {
+func validateWithCacheForAllCertificates(verifiedChains []*x509.Certificate) bool {
 	n := len(verifiedChains) - 1
-	ocspStatusChan := make(chan *ocspStatus, n)
 	for j := 0; j < n; j++ {
-		go getRevocationStatus(wg, ocspStatusChan, verifiedChains[j], verifiedChains[j+1])
+		subject := verifiedChains[j]
+		issuer := verifiedChains[j+1]
+		if valid, _, _ := validateWithCache(subject, issuer); !valid {
+			return false
+		}
 	}
-	results := make([]*ocspStatus, n)
-	for j := 0; j < n; j++ {
-		results[j] = <-ocspStatusChan // will wait for all results back
+	return true
+}
+
+func validateWithCache(subject, issuer *x509.Certificate) (bool, []byte, *certIDKey) {
+	ocspReq, err := ocsp.CreateRequest(subject, issuer, &ocsp.RequestOptions{})
+	if err != nil {
+		glog.V(2).Infof("failed to create OCSP request from the certificates.\n")
+		return false, nil, nil
 	}
-	close(ocspStatusChan)
-	return results
+	encodedCertID, ocspS := extractCertIDKeyFromRequest(ocspReq)
+	if ocspS.code != ocspSuccess {
+		glog.V(2).Infof("failed to extract CertID from OCSP Request.\n")
+		return false, ocspReq, nil
+	}
+	ocspValidatedWithCache := checkOCSPResponseCache(encodedCertID, subject, issuer)
+	if ocspValidatedWithCache.code != ocspSuccess {
+		glog.V(2).Infof("failed to validate OCSP Response. Ignored.\n")
+		return false, ocspReq, encodedCertID
+	}
+	return true, ocspReq, encodedCertID
+}
+
+func downloadOCSPCacheServer() {
+	if strings.EqualFold(os.Getenv(cacheServerEnabledEnv), "false") {
+		glog.V(2).Infof("skipping downloading OCSP Cache.")
+		return
+	}
+	ocspClient := &http.Client{
+		Timeout:   30 * time.Second,
+		Transport: snowflakeInsecureTransport,
+	}
+	ocspURL := fmt.Sprintf("%v/%v", cacheServerURL, cacheFileBaseName)
+	glog.V(2).Infof("downloading OCSP Cache from server %v", ocspURL)
+	ret := retryOCSPCacheServer(ocspClient, http.NewRequest, ocspURL, retryOCSPTimeout, retryOCSPHTTPTimeout)
+
+	ocspResponseCacheLock.Lock()
+	for k, cacheValue := range *ret {
+		if extractOCSPCacheResponseValue(cacheValue) == ocsp.Unknown {
+			continue
+		}
+		cacheKey := encodeCertIDKey(k)
+		ocspResponseCache[*cacheKey] = cacheValue
+	}
+	cacheUpdated = true
+	ocspResponseCacheLock.Unlock()
 }
 
-func getAllRevocationStatusSerial(wg *sync.WaitGroup, verifiedChains []*x509.Certificate) []*ocspStatus {
+func getAllRevocationStatus(verifiedChains []*x509.Certificate) []*ocspStatus {
+	cached := validateWithCacheForAllCertificates(verifiedChains)
+	if !cached {
+		downloadOCSPCacheServer()
+	}
```

Comment:
validate the status using cache first. If not cached, download the cache from the OCSP response cache server.